### DISCUSSION TITLE
Install AutoTs on Azure ML kernel and Use it in the Note Book for Time Series

### DISCUSSION THREAD

**Discussion by RCML2016** [winedarksea/AutoTS#170] [Q&A]
Hello All, I am trying to Use Autots for forecasting in the Azure ML Note Book.

Used below :
!pip3 install auto-ts

The Output is : /bin/bash: pip3: command not found.

Any Suggestions ?

**Comment by winedarksea** [comment]
This is nothing to do with AutoTS but rather with Python. You can do some more research, there are plenty of answers on Google.
For Linux, for example, you might run: ` sudo apt install python3-pip`

**Reply by winedarksea** [reply]
also this is `pip install autots`, auto-ts with the hypen is actually a different package

**Reply by RCML2016** [reply]
> also this is `pip install autots`, auto-ts with the hypen is actually a different package

The !pip install autots worked on Azure ML Note Book.

**Comment by RCML2016** [comment]
Hi, How to use the Holidays Package ? Is there any Pre-requisites and Pre installs  for Holidays Package. I am using AutoTs and Its Even Predicting for Weekends and Holidays. I would like the Models to be Not to future forecasting for Weekends/Holidays.

**Reply by winedarksea** [reply]
!pip install holidays from [here](https://pypi.org/project/holidays/)
that's for the holiday_country usage. The HolidayDetector/HolidayTransformer for automatic holiday detection do not require additional packages.

**Reply by winedarksea** [reply]
I don't entirely understand your question. Did you want to _not_ predict holidays?

**Reply by RCML2016** [reply]
> I don't entirely understand your question. Did you want to _not_ predict holidays?

Yes , I don't want to NO predict  for Holidays. Example Jan-01-2023 or future May-29-2023 are Holidays and some Stores will be Not Opened. So NO Prediction will help in consolidating the Over All Sales at Daily level.

**Reply by winedarksea** [reply]
`constraint` is probably your best bet for that.
For the example dataset of `df = load_daily(long=False)`
```python
forecast_index = pd.date_range(start=df.index[-1], periods=forecast_length + 1, freq=df.index.freq)[1:]
# sets an extremely high value for the cap, one that should never actually be reached by the data normally
upper_constraint = pd.DataFrame(9999999999, index=forecast_index, columns=df.columns)
# in this case also assuming negatives won't happen so setting a lower constraint of 0
lower_constraint = pd.DataFrame(0, index=forecast_index, columns=df.columns)
# add in your dates you want as definitely 0
upper_constraint.loc["2022-10-31"] = 0  # can do this to specific columns if some are opened, some closed
constraint = {
    "constraint_method": "absolute",
    "upper_constraint": upper_constraint,
    "lower_constraint": lower_constraint,
    "bounds": True,
}
model = AutoTS(
    _your args here_
)
prediction = model.predict(...)
prediction.apply_constraints(**constraint)
```
constraints can be passed to AutoTS as part of training, however the need to fit a 0 often skews the error metrics and results in a poor model choice. Better to do it after in the case of extreme constraints.

another solution is to pass in an external (`future_regressor`) which contains a flag (0 or 1 feature) for historical holidays (in training) and for future holidays for the predict.

**Comment by RCML2016** [comment]
Hello Experts, When i use the ****model_list='probabilistic' ** and Sort the fit with sMAPE, MAE, RMSE..etc ,

 I am seeing something Model Name as "Ensemble" .  Which is some thing different other than FBProphet or ETS..  What is Ensemble and Why i am seeing it in Models List ??

	Model	Runs	smape	mae	rmse	Score	TotalRuntimeSeconds
650	FBProphet	1	12.533976	288191.856204	497702.808894	24.256076	0.3352
777	Ensemble	1	12.567421	288779.309279	499049.020462	24.079424	2.0422
648	FBProphet	1	12.576267	289133.565814	499448.183847	24.175259	0.3896
775	Ensemble	1	12.607595	289769.620251	500594.426445	24.212839	3.2829
647	Ensemble	1	12.610875	289852.744161	500679.228767	24.283677	1.0805
774	Ensemble	1	12.642774	290549.878459	501678.295230	24.229837	2.5530
507	FBProphet	1	12.678483	292753.181623	506508.080901	24.246685	0.3811
406	FBProphet	1	12.762771	294555.926160	508768.922959	24.347063	0.3465
769	Ensemble	1	12.767177	294642.949796	508857.163158	24.104104	2.0248
646	Ensemble	1	12.773827	293428.072058	505590.046716	23.814324

**Reply by winedarksea** [reply]
An Ensemble is a combination of other models. It will only use the models you have listed, so in this case it will be some combination of Prophet and ETS models (you can see which in the ModelParameters).
Ensembling usually improves accuracy, but is easy enough to turn off. Set `ensemble=None` in the AutoTS() args.

**Reply by RCML2016** [reply]
> An Ensemble is a combination of other models. It will only use the models you have listed, so in this case it will be some combination of Prophet and ETS models (you can see which in the ModelParameters). Ensembling usually improves accuracy, but is easy enough to turn off. Set `ensemble=None` in the AutoTS() args.

@winedarksea , Thank You.

**Comment by RCML2016** [comment]
Getting Error : 
TypeError: infer_freq() got an unexpected keyword argument 'warn' on the Fit.

Below are my args for the Fit.

Model.fit(DF, date_col='Date', value_col='ValueCol', id_col=None)

Anyone Experiencing , I am using this on the Azure Note books. Thnaks.

**Reply by winedarksea** [reply]
This is from pandas infer_freq function. It looks like the "warn" argument has been removed in the latest release of Pandas. You must have a new environment using Pandas 2.0.0, downgrading/installing to pandas version 1.x should work
